plugins/postgres.php

```diff
@@ -0,0 +1,177 @@
+    }
+
+    private function getConnection()
+    {
+        return $this->connection ?? pg_pconnect("host=127.0.0.1 port=$this->port user=$this->user");
+    }
+
+    // attempts to fetch info about engine and return it
+    protected function getInfo()
+    {
+
+        $connection = $this->getConnection();
+        $status = pg_connection_status($connection);
+        if ($status === PGSQL_CONNECTION_OK) {
+            $result = pg_query($connection, "SELECT name, setting FROM pg_settings;");
+            if ($result) {
+                $out = [];
+                while ($row = pg_fetch_row($result)) {
+                    $out[$row[0]] = $row[1];
+                }
+                $ret['variables'] = $out;
+            }
+
+            $result = pg_query($connection, "SHOW server_version;");
+            if ($result) {
+                $result = pg_fetch_all($result);
+                $ret['version'] = $result[0]['server_version'];
+            }
+        }
+
+
+        return $ret ?? false;
+    }
+
+    protected function appendType($info): string
+    {
+        return "";
+    }
+
+    protected function prepareQuery($query)
+    {
+        return $query; // no modifications required
+    }
+
+    // returns true when it's possible to connect to engine and it can run a simple query
+    protected function canConnect(): bool
+    {
+        $connection = $this->getConnection();
+        $status = pg_connection_status($connection);
+        if ($status === PGSQL_CONNECTION_OK) {
+            $result = pg_query($connection, "show all");
+            if ($result) {
+                return true;
+            }
+
+        } elseif (pg_ping($connection)) {
+            $result = pg_query($connection, "show all");
+            if ($result) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    protected function beforeQuery():void
+    {
+        $this->connection = $this->getConnection();
+
+        $timeout = self::$commandLineArguments['query_timeout'] * 1000;
+        pg_query($this->connection, "SET statement_timeout TO ".$timeout.";");
+    }
+
+    // runs one query against engine
+    // must respect self::$commandLineArguments['query_timeout']
+    // must return ['timeout' => true] in case of timeout
+    protected function testOnce($query)
+    {
+        if (!pg_connection_busy($this->connection)) {
+            pg_send_query($this->connection, $query.';');
+        }
+
+        $res = pg_get_result($this->connection);
+        if ($res) {
+            $state = pg_result_error_field($res, PGSQL_DIAG_SQLSTATE);
+
+            if ($state === null) {
+                return $res;
+            }
+
+            $errorDescription = pg_result_error($res);
+            $out              = ['mysqlError' => trim($errorDescription), 'mysqlErrorCode' => $state];
+
+            if ($state === "57014") {
+                $out['timeout'] = true;
+            }
+
+            return $out;
```

Comment:
It seems better to use pg_query and move parsing the result to parseResult(). See https://github.com/db-benchmarks/db-benchmarks/blob/main/plugins/mysql.php#L69-L95

Reply:
As discussed and written in the [docs](https://www.php.net/manual/en/function.pg-result-error.php) it's better to use `pg_send_query()` if we want to know error state.